contracts/cw-govmod-single/src/tests.rs

```diff
@@ -0,0 +1,680 @@
+        .wrap()
+        .query_wasm_smart(govmod_single, &QueryMsg::Proposal { proposal_id: 1 })
+        .unwrap();
+
+    assert_eq!(proposal.proposal.status, expected_status)
+}
+
+#[test]
+fn test_vote_simple() {
+    do_test_votes(
+        vec![TestVote {
+            voter: "ekez".to_string(),
+            position: Vote::Yes,
+            weight: Uint128::new(10),
+            should_execute: ShouldExecute::Yes,
+        }],
+        Threshold::AbsolutePercentage {
+            percentage: Decimal::percent(100),
+        },
+        Status::Passed,
+        None,
+    );
+
+    do_test_votes(
+        vec![TestVote {
+            voter: "ekez".to_string(),
+            position: Vote::No,
+            weight: Uint128::new(10),
+            should_execute: ShouldExecute::Yes,
+        }],
+        Threshold::AbsolutePercentage {
+            percentage: Decimal::percent(100),
+        },
+        Status::Rejected,
+        None,
+    )
+}
+
+#[test]
+fn test_vote_no_overflow() {
+    // We should not overflow when computing passing thresholds even
+    // when there are 2^128 votes.
+    do_test_votes(
+        vec![TestVote {
+            voter: "ekez".to_string(),
+            position: Vote::Yes,
+            weight: Uint128::new(u128::max_value()),
+            should_execute: ShouldExecute::Yes,
+        }],
+        Threshold::AbsolutePercentage {
+            percentage: Decimal::percent(100),
+        },
+        Status::Passed,
+        None,
+    );
+}
+
+#[test]
+fn test_vote_abstain_only() {
+    do_test_votes(
+        vec![TestVote {
+            voter: "ekez".to_string(),
+            position: Vote::Abstain,
+            weight: Uint128::new(u128::max_value()),
+            should_execute: ShouldExecute::Yes,
+        }],
+        Threshold::AbsolutePercentage {
+            percentage: Decimal::percent(100),
+        },
+        Status::Open,
+        None,
+    );
+}
+
+#[test]
+fn test_no_double_votes() {
+    do_test_votes(
+        vec![
+            TestVote {
+                voter: "ekez".to_string(),
+                position: Vote::Abstain,
+                weight: Uint128::new(2),
+                should_execute: ShouldExecute::Yes,
+            },
+            TestVote {
+                voter: "ekez".to_string(),
+                position: Vote::Yes,
+                weight: Uint128::new(2),
+                should_execute: ShouldExecute::No,
+            },
+        ],
+        Threshold::AbsolutePercentage {
+            percentage: Decimal::percent(100),
+        },
+        Status::Open,
+        None,
+    );
+}
+
+/// The current behavior for passing proposals is that the first
+/// option to reach the threshold wins. For example, with a 50%
+/// passing threshold if 50% of voting power votes no that proposal
+/// fails even if the other 50% would have voted yes. The same goes if
+/// the yes and no were reversed.
+///
+/// TODO(zeke): is this the behavior that we want?
```

Comment:
I would vote that if it's still possible for the proposal to pass, it should not be rejected